lib/aws/client.ex

```diff
@@ -195,11 +195,61 @@ defmodule AWS.Client do
   end
 
   def request(client, method, url, body, headers, opts \\ []) do
+    {retry_num, opts} = Keyword.pop(opts, :retry_num, 0)
+    {retry_opts, opts} = Keyword.pop(opts, :retry_opts, max_retries: 3)
+
     {mod, options} = Map.fetch!(client, :http_client)
     options = Keyword.merge(options, opts)
-    apply(mod, :request, [method, url, body, headers, options])
+    resp = apply(mod, :request, [method, url, body, headers, options])
+
+    retriable?(resp)
+    |> case do
+      :ok ->
+        resp
+
+      :retry ->
+        if should_retry?(retry_num, retry_opts) do
+          updated_opts = Keyword.merge(opts, retry_num: retry_num + 1, retry_opts: retry_opts)
+          request(client, method, url, body, headers, updated_opts)
+        else
+          resp
+        end
+
+      :error ->
+        resp
+    end
+  end
+
+  def should_retry?(_retry_num, nil), do: false
```

Comment:
Since you have a default on line 199 this can never be `nil` right? (It'll default to: `[max_retries: 3]`)

Reply:
Yikes.. you are absolutely right. I fumble-committed that by a mistake. Good catch!